exercises/src/main/scala/fpinscala/gettingstarted/GettingStarted.scala

```diff
@@ -140,7 +146,11 @@ object PolymorphicFunctions {
 
   // Exercise 2: Implement a polymorphic function to check whether
   // an `Array[A]` is sorted
-  def isSorted[A](as: Array[A], gt: (A,A) => Boolean): Boolean = ???
+  def isSorted[A](as: Array[A], gt: (A,A) => Boolean): Boolean = {
+    if (as.length < 2) true
+    else if (gt(as(0), as(1))) false
+    else isSorted(as.tail, gt)
+  }
```

Comment:
Your implementation will work, but from a performance point of view, you would generally not copy an array by calling tail, it's not an efficient operation like it is on `List[A]`.  You could either convert the array to a list, and then se the sort of recursion you're doing, although, in this case I would avoid using `as.tail` for the same reason as I was saying yesterday about `as.head`.

But what I believe they are going for is that you should have in inner loop function like `go` in `binarySearch` above, and pass indexes around and always access the same shared `Array[A]`.

Reply:
I didn't test the first implementation to confirm it was working, but the second one isn't correct:

```
> exercises/console
[info] Starting scala interpreter...
[info]
Welcome to Scala 2.12.1 (OpenJDK 64-Bit Server VM, Java 1.8.0_252).
Type in expressions for evaluation. Or try :help.

scala> import fpinscala.gettingstarted.PolymorphicFunctions._
import fpinscala.gettingstarted.PolymorphicFunctions._

scala> val array = Array(1, 2, 5, 10)
array: Array[Int] = Array(1, 2, 5, 10)

scala> isSorted[Int](array, _ > _)
java.lang.ArrayIndexOutOfBoundsException: 4
  at scala.runtime.ScalaRunTime$.array_apply(ScalaRunTime.scala:55)
  at fpinscala.gettingstarted.PolymorphicFunctions$.go$4(GettingStarted.scala:154)
  at fpinscala.gettingstarted.PolymorphicFunctions$.isSorted(GettingStarted.scala:157)
  ... 39 elided

scala> isSorted[Int](array, _ < _)
res1: Boolean = false
```